middleware/app/Utilities/CanvasUtil.php

```diff
@@ -0,0 +1,1275 @@
+<?php
+
+declare(strict_types=1);
+
+use GuzzleHttp\Client;
+use GuzzleHttp\Exception\RequestException;
+use App\Models\ProviderPlatform;
+
+
+class CanvasUtil extends ProviderUtil
+{
+
+    public function __construct(string $type, int $accountId, string $accountName, string $apiKey, string $url, string $iconUrl)
+    {
+        return parent::__construct($type, $accountId,  $accountName, $apiKey, $url, $iconUrl);
+    }
+
+
+    // constructor for when we already have the providerId
+    public static function getByProviderId($providerId): CanvasUtil | \InvalidArgumentException
+    {
+        $provider = ProviderPlatform::findByProviderId($providerId);
+
+        if (!$provider) {
+            throw new \InvalidArgumentException('Invalid provider ID');
+        }
+        return new self($provider->type, $provider->account_id, $provider->account_name, $provider->access_key, $provider->base_url, $provider->icon_url);
+    }
+
+    /**
+     * Validate and format the account ID parameter for API URLs
+     *
+     * @param string $id
+     * @return string Formatted account or user ID
+     * @throws \InvalidArgumentException If the account ID is invalid
+     */
+    public function fmtAndValidateId(string $id): string
+    {
+        if ($id === 'self' || is_numeric($id)) {
+            // Append a trailing slash if needed
+            if (substr($id, -1) !== '/') {
+                $id .= '/';
+            }
+            return $id;
+        } else {
+            throw new \InvalidArgumentException('Invalid account ID');
+        }
+    }
+
+    /**
+     * Get a list of users from Canvas
+     *
+     * @param? string
+     * @return mixed JSON decoded
+     * @throws \Exception
+     *
+     * AccountId can be accessed via the field in the class,
+     * but it seems most of the time it will be self.
+     */
+    public function listUsers(string $accountId = 'self'): mixed
+    {
+        $base_url = $this->base_url . "api/v1/";
+        $access_key = $this->access_key;
+
+        $client = new Client([
+            'headers' => [
+                'Authorization' => 'Bearer ' . $access_key,
+            ],
+        ]);
+        $accountId = self::fmtAndValidateId($accountId);
```

Comment:
In the future, we might want to abstract out all the boilerplate from each of these.